diff --git a/website/macros.bzl b/website/macros.bzl
index ff93e2d74..1e7cbd547 100644
--- a/website/macros.bzl
+++ b/website/macros.bzl
@@ -96,7 +96,7 @@ def static_website(
             TAR_COMMAND=$$(which tar)
         fi
 
-        $$TAR_COMMAND -xf $$SOURCE $${DECOMPRESS_ARGS:+DECOMPRESS_ARGS}
+        $$TAR_COMMAND -xf $$SOURCE "$$DECOMPRESS_ARGS"
 
         while IFS= read -r CMD; do
             $$CMD
