diff --git a/api/bazel/repository_locations.bzl b/api/bazel/repository_locations.bzl
index 3ca6722102..7504cc7849 100644
--- a/api/bazel/repository_locations.bzl
+++ b/api/bazel/repository_locations.bzl
@@ -155,8 +155,8 @@ REPOSITORY_LOCATIONS_SPEC = dict(
         project_name = "envoy_toolshed",
         project_desc = "Tooling, libraries, runners and checkers for Envoy proxy's CI",
         project_url = "https://github.com/envoyproxy/toolshed",
-        version = "0.0.9",
-        sha256 = "f1a2169d271efbf4de2b24207136c5009c5453fba62ba8dc3497cba204d092aa",
+        version = "0.0.10",
+        sha256 = "bdfcf0a23c18a99887ac25761aa56d85bedb6eda77c89f9f19e6142b812749b9",
         strip_prefix = "toolshed-bazel-v{version}/bazel",
         urls = ["https://github.com/envoyproxy/toolshed/archive/bazel-v{version}.tar.gz"],
         use_category = ["build"],
diff --git a/ci/do_ci.sh b/ci/do_ci.sh
index e63cc80be0..c3a6360a7b 100755
--- a/ci/do_ci.sh
+++ b/ci/do_ci.sh
@@ -648,8 +648,6 @@ case $CI_TARGET in
         setup_clang_toolchain
         echo "generating docs..."
         # Build docs.
-        # We want the binary at the end
-        BAZEL_BUILD_OPTIONS+=(--remote_download_toplevel)
         [[ -z "${DOCS_OUTPUT_DIR}" ]] && DOCS_OUTPUT_DIR=generated/docs
         rm -rf "${DOCS_OUTPUT_DIR}"
         mkdir -p "${DOCS_OUTPUT_DIR}"
@@ -661,8 +659,12 @@ case $CI_TARGET in
             bazel "${BAZEL_STARTUP_OPTIONS[@]}" build "${BAZEL_BUILD_OPTIONS[@]}" //docs:rst
             cp bazel-bin/docs/rst.tar.gz "$DOCS_OUTPUT_DIR"/envoy-docs-rst.tar.gz
         fi
-        bazel "${BAZEL_STARTUP_OPTIONS[@]}" build "${BAZEL_BUILD_OPTIONS[@]}" //docs:html
-        tar -xzf bazel-bin/docs/html.tar.gz -C "$DOCS_OUTPUT_DIR"
+        DOCS_OUTPUT_DIR="$(realpath "$DOCS_OUTPUT_DIR")"
+        bazel "${BAZEL_STARTUP_OPTIONS[@]}" run \
+              "${BAZEL_BUILD_OPTIONS[@]}" \
+              --//tools/tarball:target=//docs:html \
+              //tools/tarball:unpack \
+              "$DOCS_OUTPUT_DIR"
         ;;
 
     docs-upload)
diff --git a/tools/tarball/BUILD b/tools/tarball/BUILD
new file mode 100644
index 0000000000..190817557e
--- /dev/null
+++ b/tools/tarball/BUILD
@@ -0,0 +1,6 @@
+load("@envoy_toolshed//tarball:macros.bzl", "unpacker")
+
+unpacker(
+    name = "unpack",
+    zstd = "//tools/zstd",
+)
