diff --git a/tools/base/entry_point.py b/tools/base/entry_point.py
index cdd1cf421d..97b566ae41 100644
--- a/tools/base/entry_point.py
+++ b/tools/base/entry_point.py
@@ -26,6 +26,24 @@ def main(*args) -> int:
         return 1
 
     try:
+        import importlib
+        import os
+        import pathlib
+        for path in sys.path:
+            path = pathlib.Path(path)
+
+            if "utils" in path.name:
+                sys.path += [str(path.joinpath("site-packages"))]
+
+            if path.name == "envoy":
+                if "tools" not in os.listdir(path):
+                    continue
+                init_file = path.joinpath("__init__.py")
+                # if not init_file.read_text():
+                init_file.write_text(
+                    "__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+                path.joinpath("tools", "__init__.py").write_text(
+                    "__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
         return subprocess.run([tool_path, *args]).returncode
     except KeyboardInterrupt:
         return 1
diff --git a/tools/docs/generate_extensions_security_rst.py b/tools/docs/generate_extensions_security_rst.py
index f28ee0347f..eb234184a1 100644
--- a/tools/docs/generate_extensions_security_rst.py
+++ b/tools/docs/generate_extensions_security_rst.py
@@ -8,6 +8,26 @@ import pathlib
 import sys
 import tarfile
 
+# PATHHACK
+import importlib
+import os
+import pathlib
+
+for path in sys.path:
+    path = pathlib.Path(path)
+
+    if "utils" in path.name:
+        sys.path += [str(path.joinpath("site-packages"))]
+
+    if path.name == "envoy":
+        if "tools" not in os.listdir(path):
+            continue
+        init_file = path.joinpath("__init__.py")
+        #if not init_file.read_text():
+        init_file.write_text("__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+        path.joinpath("tools", "__init__.py").write_text(
+            "__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+
 from envoy.base import utils
 
 
diff --git a/tools/docs/generate_version_histories.py b/tools/docs/generate_version_histories.py
index 6bbb16f34b..a04adb756b 100644
--- a/tools/docs/generate_version_histories.py
+++ b/tools/docs/generate_version_histories.py
@@ -10,6 +10,26 @@ from packaging import version as _version
 
 from aio.run import runner
 
+# PATHHACK
+import importlib
+import os
+import pathlib
+
+for path in sys.path:
+    path = pathlib.Path(path)
+
+    if "utils" in path.name:
+        sys.path += [str(path.joinpath("site-packages"))]
+
+    if path.name == "envoy":
+        if "tools" not in os.listdir(path):
+            continue
+        init_file = path.joinpath("__init__.py")
+        #if not init_file.read_text():
+        init_file.write_text("__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+        path.joinpath("tools", "__init__.py").write_text(
+            "__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+
 from envoy.base import utils
 from envoy.base.utils import IProject, Project
 
diff --git a/tools/protodoc/manifest_to_json.py b/tools/protodoc/manifest_to_json.py
index 21defb0d0a..39da6e1567 100644
--- a/tools/protodoc/manifest_to_json.py
+++ b/tools/protodoc/manifest_to_json.py
@@ -4,6 +4,26 @@ import sys
 
 from google.protobuf import json_format
 
+# PATHHACK
+import importlib
+import os
+import pathlib
+
+for path in sys.path:
+    path = pathlib.Path(path)
+
+    if "utils" in path.name:
+        sys.path += [str(path.joinpath("site-packages"))]
+
+    if path.name == "envoy":
+        if "tools" not in os.listdir(path):
+            continue
+        init_file = path.joinpath("__init__.py")
+        #if not init_file.read_text():
+        init_file.write_text("__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+        path.joinpath("tools", "__init__.py").write_text(
+            "__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+
 from envoy.base.utils import ProtobufValidator
 
 from tools.protodoc import manifest_pb2
diff --git a/tools/protodoc/protodoc.py b/tools/protodoc/protodoc.py
index 88ac1f7b66..0535175856 100755
--- a/tools/protodoc/protodoc.py
+++ b/tools/protodoc/protodoc.py
@@ -19,6 +19,26 @@ from udpa.annotations import status_pb2 as udpa_status_pb2
 from validate import validate_pb2
 from xds.annotations.v3 import status_pb2 as xds_status_pb2
 
+# PATHHACK
+import importlib
+import os
+import pathlib
+
+for path in sys.path:
+    path = pathlib.Path(path)
+
+    if "utils" in path.name:
+        sys.path += [str(path.joinpath("site-packages"))]
+
+    if path.name == "envoy":
+        if "tools" not in os.listdir(path):
+            continue
+        init_file = path.joinpath("__init__.py")
+        #if not init_file.read_text():
+        init_file.write_text("__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+        path.joinpath("tools", "__init__.py").write_text(
+            "__path__ = __import__('pkgutil').extend_path(__path__, __name__)")
+
 from tools.api_proto_plugin import annotations, constants, plugin, visitor
 from tools.protodoc import jinja
 from tools.protodoc.data import data
